src/streaming/zarrv2.array.writer.cpp

```diff
@@ -0,0 +1,177 @@
+#include "macros.hh"
+#include "zarrv2.array.writer.hh"
+#include "sink.creator.hh"
+#include "zarr.common.hh"
+
+#include <nlohmann/json.hpp>
+
+#include <latch>
+#include <stdexcept>
+
+namespace {
+std::string
+sample_type_to_dtype(ZarrDataType t)
+
+{
+    const std::string dtype_prefix =
+      std::endian::native == std::endian::big ? ">" : "<";
+
+    switch (t) {
+        case ZarrDataType_uint8:
+            return dtype_prefix + "u1";
+        case ZarrDataType_uint16:
+            return dtype_prefix + "u2";
+        case ZarrDataType_uint32:
+            return dtype_prefix + "u4";
+        case ZarrDataType_uint64:
+            return dtype_prefix + "u8";
+        case ZarrDataType_int8:
+            return dtype_prefix + "i1";
+        case ZarrDataType_int16:
+            return dtype_prefix + "i2";
+        case ZarrDataType_int32:
+            return dtype_prefix + "i4";
+        case ZarrDataType_int64:
+            return dtype_prefix + "i8";
+        case ZarrDataType_float32:
+            return dtype_prefix + "f4";
+        case ZarrDataType_float64:
+            return dtype_prefix + "f8";
+        default:
+            throw std::runtime_error("Invalid data type: " +
+                                     std::to_string(static_cast<int>(t)));
+    }
+}
+} // namespace
+
+zarr::ZarrV2ArrayWriter::ZarrV2ArrayWriter(
+  const ArrayWriterConfig& config,
+  std::shared_ptr<ThreadPool> thread_pool,
+  std::shared_ptr<S3ConnectionPool> s3_connection_pool)
+  : ArrayWriter(config, thread_pool, s3_connection_pool)
+{
+}
+
+zarr::ZarrV2ArrayWriter::~ZarrV2ArrayWriter()
+{
+    is_finalizing_ = true;
```

Comment:
I've mentioned this before, but I think that having code that can fail in the destructor is actively bad. Destructors should not fail, throw exceptions, or log errors because there's no way to handle these issues gracefully—you don't know exactly how or when they might be executed. There's also a potential for memory leaks, and no explicit control over resource management. This approach can lead to unpredictable behavior and difficult-to-debug issues.